Mathlib/RingTheory/DividedPowers/Basic.lean

```diff
@@ -0,0 +1,514 @@
+    by_cases hn : n = 0
+    · rw [if_pos hn, hn, if_pos rfl, _root_.pow_zero]
+    · simp only [if_neg hn]
+  dpow_mul m n {x} hx := by
+    rw [Ideal.mem_bot.mp hx]
+    simp only [true_and, mul_ite, mul_one, mul_zero, add_eq_zero]
+    by_cases hn : n = 0
+    · simp only [hn, ite_true, and_true, add_zero, choose_self, cast_one]
+    · rw [if_neg hn, if_neg]
+      exact not_and_of_not_right (m = 0) hn
+  dpow_comp m {n a} hn ha := by
+    rw [Ideal.mem_bot.mp ha]
+    simp only [true_and, ite_eq_right_iff, _root_.mul_eq_zero, mul_ite, mul_one, mul_zero]
+    by_cases hm: m = 0
+    · simp only [hm, and_true, true_or, ite_true, uniformBell_zero_left, cast_one]
+      rw [if_pos]
+      exact fun h => False.elim (hn h)
+    · simp only [hm, and_false, ite_false, false_or, if_neg hn]
+
+instance {A : Type*} [CommSemiring A] [DecidableEq A] :
+    Inhabited (DividedPowers (⊥ : Ideal A)) := ⟨dividedPowersBot A⟩
+
+/-- The coercion from the divided powers structures to functions -/
+instance {A : Type*} [CommSemiring A] (I : Ideal A) :
+    CoeFun (DividedPowers I) fun _ => ℕ → A → A := ⟨fun hI => hI.dpow⟩
+
+/- -- The synTaut linter complains about this theorem.
+theorem coe_apply {A : Type*} [CommSemiring A]
+    (I : Ideal A) (hI : DividedPowers I) (n : ℕ) (a : A) :
+  hI n a = hI.dpow n a := rfl -/
+
+theorem coe_injective {A : Type*} [CommSemiring A] (I : Ideal A) :
+    Function.Injective (fun (h : DividedPowers I) ↦ (h : ℕ → A → A)) := fun hI hI' h ↦ by
+  ext n x
+  exact congr_fun (congr_fun h n) x
+
+@[ext]
+theorem eq_of_eq_on_ideal {A : Type*} [CommSemiring A]
+    {I : Ideal A} (hI : DividedPowers I) (hI' : DividedPowers I)
+    (h_eq : ∀ (n : ℕ) {x : A} (_ : x ∈ I), hI.dpow n x = hI'.dpow n x) :
+    hI = hI' := by
+  ext n x
+  by_cases hx : x ∈ I
+  · exact h_eq n hx
+  · rw [hI.dpow_null hx, hI'.dpow_null hx]
+
+/- structure dpRing (A : Type*) extends CommSemiring A where
+  dpIdeal : Ideal A
+  dividedPowers : DividedPowers dpIdeal
+
+instance {A : Type*} [CommSemiring A] [DecidableEq A] : Inhabited (dpRing A)
+  where default :=
+  { toCommSemiring := inferInstance
+    dpIdeal := ⊥
+    dividedPowers := dividedPowersBot A }
+-/
+
+end DividedPowersDefinition
+
+namespace DividedPowers
+
+open Finset Nat
+
+section BasicLemmas
+
+/- ## Basic lemmas for divided powers -/
+
+variable {A : Type*} [CommSemiring A] {I : Ideal A}
+
+/-- Variant of `DividedPowers.dpow_add` with a sum on `range (n + 1)` -/
+theorem dpow_add' (hI : DividedPowers I) (n : ℕ) {x y : A} (hx : x ∈ I) (hy : y ∈ I) :
+    hI.dpow n (x + y) = (range (n + 1)).sum fun k => hI.dpow k x * hI.dpow (n - k) y := by
+  rw [hI.dpow_add n hx hy, sum_antidiagonal_eq_sum_range_succ_mk]
+
+/-- The exponential series of an element in the context of divided powers,
+`Σ (dpow n a) X ^ n` -/
+def exp (hI : DividedPowers I) (a : A) :=
+  PowerSeries.mk fun n => hI.dpow n a
+
+theorem exp_add_aux (dp : ℕ → A → A) {a b : A} -- (ha : a ∈ I) (hb : b ∈ I) :
```

Comment:
Do you need this version?

Reply:
I'd say yes, for future work, when you wish to get `dp_add` from the multiplicativity property of the exponential series only.